cmd/terramate/cli/cli.go

```diff
@@ -360,15 +361,32 @@ func (c *cli) generateGraph() error {
 
 	loader := stack.NewLoader(c.root())
 	di := dot.NewGraph(dot.Directed)
+	d := dag.New()
 
+	visited := map[string]struct{}{}
 	for _, e := range c.filterStacksByWorkingDir(entries) {
-		tree, err := terramate.BuildOrderTree(c.root(), e.Stack, loader)
+		if _, ok := visited[e.Stack.Dir]; ok {
+			continue
+		}
+
+		err := terramate.BuildDAG(d, c.root(), e.Stack, loader, visited)
 		if err != nil {
 			return fmt.Errorf("failed to build order tree: %w", err)
 		}
+	}
+
+	for _, id := range d.IDs() {
+		val, err := d.Node(id)
+		if err != nil {
+			return fmt.Errorf("generating graph: %w", err)
+		}
```

Comment:
The problem goes deeper than that =)

The dag can have intermediate states where some nodes don't have values, because we build a single dag so the `before` entries of a node creates `after` entries in their own nodes but at this point, their value (the stack.S object) wasn't loaded yet. We cannot switch to fully-evaluated nodes with values because so we hit the cycle before the DAG is built.

I tried to avoid this concept of `Node` type since the beginning for the internal implementation, but no problem exposing it to users, but if we do it, the `node.Value()` can still return nil and we will need to handle this case in the same way.

So eventually all stacks are loaded and the method `dag.AddNode()` is called for each one (remember that for some of them, their d.dag[id] is already built from the `before` clause of other nodes) and then the DAG is "complete". My initial idea for the `dag.Validate()` was to validate also the completeness of the DAG, but I actually forgot to do that ... 

I see some ways of improving this but I'd like to avoid the mix of DAG and stack loading logic together (as in my first implementation for `after`).

Reply:
Again my intuition is very naive (to not say stupid :laughing:), but it seems that maybe splitting DAG building and the DAG itself could make these problems more explicit (and the DAG would always be complete/have nodes/dont have cycles). Like the wholw building/validation.etc logic happens on DAGBuilder thing, and when you are done (and validated), then you get an actual DAG (and nodes always have values). Something like:

```go
builder := DAGBuilder()
builder.add()
builder.add()
//so goes on

dag := builder.Done() // would be the validate
//now you have a full DAG, no cycles, just fun and values
```

It is just my intuition against objects with intermediary state that is invalid (hence why we have Validate), something like a builder makes this more explicit because you are still just building something, once you have a DAG (or pass one as a parameter), for sure it is complete and correct, etc. But honestly don't know if it makes any sense, just sharing, maybe it will help with some idea, maybe it is the stupidiest shit ever :laughing: